### PR TITLE
feat: use environment in paths

### LINKED ISSUE #2935: Rework auth in YDB EM internal installation
Use different session services depending on cluster name

### PR DESCRIPTION
**Pull request by Raubzeug**
closes #2935
[Stand](https://nda.ya.ru/t/YvL8JQvv7M4DRn) - only this one cluster is supported for now.

<h2>Greptile Overview</h2>

Updated On: 2025-10-28 10:47:53 UTC

<h3>Greptile Summary</h3>

This PR introduces environment-based routing to support different authentication services based on cluster environment. The implementation adds an optional `environment` parameter to all major routes (`/:environment?/cluster`, `/:environment?/tenant`, etc.) and extracts the environment from the URL's first path segment.

**Key Changes:**
- Routes now support optional environment prefix (e.g., `/cloud-prod/cluster`)
- Environment is extracted from URL when it matches the `allowedEnvironments` list
- Created React hooks (`useVDiskPagePath`, `useStorageGroupPath`, `useTabletPagePath`) to ensure database context is properly included in generated paths
- API calls prepend environment prefix to meta proxy paths
- Refactored path generation across 69 files for consistency

**Architecture:**
- Environment extraction happens at app initialization in `getUrlData`
- Environment is stored globally and used in `createHref` to maintain consistent routing
- Components use hooks instead of direct path generation to avoid prop drilling

The changes are well-structured and include comprehensive test coverage for environment extraction logic.

<h3>Confidence Score: 4/5</h3>

- This PR is safe to merge with minor risk related to potential edge cases in route matching
- The implementation is well-structured with comprehensive test coverage for URL parsing logic. The changes follow consistent patterns across the codebase. However, the extensive refactoring (69 files) and introduction of optional route parameters could have edge cases. The use of hooks for path generation is a good architectural choice that avoids prop drilling.
- Pay close attention to `src/routes.ts` for route matching logic with optional environment parameter

<h3>Important Files Changed</h3>

File Analysis

| Filename | Score | Overview |
|----------|-------|----------|
| src/routes.ts | 4/5 | Added environment parameter to route definitions and created hooks (useVDiskPagePath, useStorageGroupPath, useTabletPagePath) to ensure database context is properly included in paths |
| src/store/getUrlData.ts | 5/5 | Extracts environment from first URL segment when in allowedEnvironments list for multi-cluster mode |
| src/store/configureStore.ts | 5/5 | Passes environments array to getUrlData and exposes environment to window object |
| src/services/api/meta.ts | 5/5 | Prepends environment prefix to meta proxy paths when environment is defined |
| src/utils/parseBalancer.ts | 5/5 | Prepends environment prefix when building backend URL from balancer and meta_backend |
| src/components/VDiskPopup/VDiskPopup.tsx | 4/5 | Refactored to use useVDiskPagePath hook instead of passing database query parameter directly |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant App as App Init
    participant URL as getUrlData
    participant Store as configureStore
    participant Routes as Route Handler
    participant API as MetaAPI
    participant Component as React Component

    App->>URL: Extract URL data
    Note over URL: Parse pathname<br/>Check first segment<br/>against allowedEnvironments
    URL-->>Store: {environment, basename, backend, clusterName}
    Store->>Store: Set global environment variable
    Store->>Store: window.environment = environment
    
    Note over Component: User navigates to page
    Component->>Component: useVDiskPagePath() hook
    Note over Component: Hook captures database<br/>from query params
    Component->>Routes: createHref(route, params, query)
    Routes->>Routes: Add environment to params if set
    Routes->>Routes: Add database to query
    Routes-->>Component: /environment/vDisk?nodeId=X&vDiskId=Y&database=Z
    
    Component->>API: API request with path
    API->>API: getPath(path, clusterName)
    Note over API: Prepend environment<br/>if proxyMeta enabled
    API-->>API: /environment/api/meta3/proxy/cluster/name/path
    API->>API: Make HTTP request
    API-->>Component: Response data
```

### Test Status: <span style="color: orange;">⚠️ FLAKY</span>
  📊 [Full Report](https://ydb-platform.github.io/ydb-embedded-ui/3020/)

  | Total | Passed | Failed | Flaky | Skipped |
  |:-----:|:------:|:------:|:-----:|:-------:|
  | 378 | 374 | 0 | 2 | 2 |

  
  <details>
  <summary>Test Changes Summary ⏭️2 </summary>

  #### ⏭️ Skipped Tests (2)
1. Scroll to row, get shareable link, navigate to URL and verify row is scrolled into view (tenant/diagnostics/tabs/queries.test.ts)
2. Copy result button copies to clipboard (tenant/queryEditor/queryEditor.test.ts)

  </details>

  ### Bundle Size: 🔺
  Current: 47.08 MB | Main: 47.08 MB
  Diff: +6.20 KB (0.01%)

  ⚠️ Bundle size increased. Please review.

  <details>
  <summary>ℹ️ CI Information</summary>

  - Test recordings for failed tests are available in the full report.
  - Bundle size is measured for the entire 'dist' directory.
  - 📊 indicates links to detailed reports.
  - 🔺 indicates increase, 🔽 decrease, and ✅ no change in bundle size.
  </details>

### CODE REVIEW

File: src/routes.ts [hunk 1]
```diff
@@ -78,13 +86,18 @@ export function createHref(
         extendedQuery = {...extendedQuery, clusterName};
     }
 
+    //if {environment: undefinded} in params - it meant we want to reset it
```
Comment (artemmufazalov) [inline src/routes.ts]: Typo - "undefinded"

File: src/utils/parseBalancer.ts [hunk 2]
```diff
@@ -65,7 +67,16 @@ export function prepareBackendFromBalancer(rawBalancer: string) {
 
     // Use meta_backend if it is defined to form backend url
     if (window.meta_backend) {
-        return normalizePathSlashes(`${window.meta_backend}/${preparedBalancer}`);
+        const metaBackend = window.meta_backend;
+        const envPrefix = environment ? `/${environment}` : '';
+
+        // If meta_backend is a full URL (has protocol), don't add leading slash
```
Comment (artemmufazalov) [inline src/utils/parseBalancer.ts]: Please, update comment, it should point on the lack of env, not the leading slash

File: src/store/getUrlData.ts [hunk 3]
```diff
@@ -28,18 +30,26 @@ export const getUrlData = ({
     if (!singleClusterMode) {
         // Multi-cluster version
         // Cluster and backend are determined by url params
+        // Extract environment from the first path segment if it's in allowedEnvironments list
+        // e.g., /cloud-preprod/api/meta3/proxy/cluster/pre-prod_global/monitoring/cluster -> environment: 'cloud-preprod'
+        // e.g., /cloud-prod/cluster -> environment: 'cloud-prod'
+        // Environment is only extracted if allowedEnvironments list is provided
```
Comment (artemmufazalov) [inline src/store/getUrlData.ts]: Explain please, why you decided to use it with `allowedEnvironments`?

**Review by artemmufazalov — CHANGES_REQUESTED** [review]

File: src/store/getUrlData.ts [hunk 4]
```diff
@@ -28,18 +30,26 @@ export const getUrlData = ({
...
+        const firstSegment = window.location.pathname.split('/').filter(Boolean)[0];
+        const environment = allowedEnvironments?.includes(firstSegment) ? firstSegment : undefined;
```
[the remark this answers is not in the document]
Comment (Raubzeug) [reply src/store/getUrlData.ts]: For now it is supposed to work only with node, only in y-t installation

**Review by artemmufazalov — DISMISSED** [review]
